src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+  static const char OCTAL_INTEGER            ='o';
+  static const char HEXADECIMAL_INTEGER      ='x';
+  static const char HEXADECIMAL_INTEGER_UPPER='X';
+  static const char SCIENTIFIC               ='e';
+  static const char SCIENTIFIC_UPPER         ='E';
+  static const char GENERAL                  ='g';
+  static const char GENERAL_UPPER            ='G';
+  static const char DECIMAL_FLOAT            ='f';
+  static const char HEXADECIMAL_FLOAT        ='a';
+  static const char HEXADECIMAL_FLOAT_UPPER  ='A';
+  static const char CHARACTER                ='c';
+  static const char CHARACTER_UPPER     ='C';
+  static const char DATE_TIME           ='t';
+  static const char DATE_TIME_UPPER     ='T';
+  static const char BOOLEAN             ='b';
+  static const char BOOLEAN_UPPER       ='B';
+  static const char STRING              ='s';
+  static const char STRING_UPPER        ='S';
+  static const char HASHCODE            ='h';
+  static const char HASHCODE_UPPER      ='H';
+  static const char LINE_SEPARATOR      ='n';
+  static const char PERCENT_SIGN        ='%';
+
+  format_specifiert() { }
+
+  format_specifiert(
+    int _index,
+    std::string _flag,
+    int _width,
+    int _precision,
+    bool _dt,
+    char c):
+      index(_index),
+      flag(_flag),
+      width(_width),
+      precision(_precision),
+      conversion(c)
+  { }
+
+  int index=-1;
+  std::string flag;
+  int width;
+  int precision;
+  bool dt=false;
+  char conversion;
+};
+
+// Format text represent a constant part of a format string.
+class format_textt
+{
+public:
+  explicit format_textt(std::string _content): content(_content) { }
+
+  explicit format_textt(const format_textt &fs): content(fs.content) { }
+
+  std::string get_content() const
+  {
+    return content;
+  }
+
+private:
+  std::string content;
+};
+
+// A format element is either a specifier or text.
+class format_elementt
+{
+public:
+  typedef enum {SPECIFIER, TEXT} format_typet;
+
+  explicit format_elementt(format_typet _type): type(_type)
+  {
+  }
+
+  explicit format_elementt(std::string s): type(TEXT)
+  {
+    fstring.emplace_back(s);
+  }
+
+  explicit format_elementt(string_constraint_generatort::format_specifiert fs):
+    type(SPECIFIER)
+  {
+    fspec.push_back(fs);
+  }
+
+  bool is_format_specifier() const
+  {
+    return type==SPECIFIER;
+  }
+
+  bool is_format_text() const
+  {
+    return type==TEXT;
+  }
+
+  string_constraint_generatort::format_specifiert get_format_specifier() const
+  {
+    PRECONDITION(is_format_specifier());
+    return fspec.back();
+  }
+
+  format_textt &get_format_text()
+  {
+    PRECONDITION(is_format_text());
+    return fstring.back();
+  }
+
+  const format_textt &get_format_text() const
+  {
+    PRECONDITION(is_format_text());
+    return fstring.back();
+  }
+
+  ~format_elementt() {}
+
+private:
+  format_typet type;
+  std::list<format_textt> fstring;
+  std::list<string_constraint_generatort::format_specifiert> fspec;
```

Comment:
Abuse of lists to just to hold one element, maybe use a one element array (with the necessary refactoring in the other functions).